site/server.ts

```
@@ -29,26 +29,28 @@ app.prepare().then(() => {
-                    if (name === "cache-control" || name === "Cache-Control") {
-                        // Next.js applies a default cache-control header of "public, max-age=0" for assets in the public/ folder
-                        // (see https://nextjs.org/docs/app/building-your-application/optimizing/static-assets#caching).
-                        // We want to cache these assets for 15 minutes.
-                        return originalSetHeader.call(this, name, "public, max-age=900");
-                    }
-                    return originalSetHeader.call(this, name, value);
-                };
-            }
+            const originalWriteHead = res.writeHead;
+            res.writeHead = function (statusCode: number, ...args: unknown[]) {
+                // since writeHead is a callback function, it's called after handle() -> we get the actual response statusCode
+                if (statusCode >= 400) {
+                    // prevent caching of error responses
+                    res.setHeader("Cache-Control", "private, no-cache, no-store, max-age=0, must-revalidate");
+                } else if (
+                    parsedUrl.pathname?.startsWith("/assets/") ||
+                    parsedUrl.pathname === "/favicon.ico" ||
+                    parsedUrl.pathname === "/apple-icon.png" ||
+                    parsedUrl.pathname === "/icon.svg" ||
+                    parsedUrl.pathname === "/robots.txt" ||
+                    parsedUrl.pathname === "/sitemap.xml"
+                ) {
+                    res.setHeader("Cache-Control", "public, max-age=900");
+                }
+
+                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                // @ts-ignore -> TS complains about passing unknown[] to writeHead, however explicit typing is complicated because writeHead can be overloaded and has two signatures
+                return originalWriteHead.apply(this, [statusCode, ...args]);
```

Passing the args as unknown[] to originalWriteHead causes the typescript error `Type 'unknown' is not assignable to type 'OutgoingHttpHeaders | OutgoingHttpHeader[] | undefined'.`. This only happens in the starter, it didn't happen in Demo.

Since `writeHead` can be overloaded and has two signatures, correct typing is hard. So I opted for just ignoring the error here.

---

I'd use `@ts-expect-error` instead to remove the ESLint disable.

---

> correct typing is hard

can't you use any? (@ts-ingore should always be just a temporary quick workaround)

---

I tried this, any doesn't work 🤷🏼‍♂️ 

Casting the arguments to one of the correct overloads works:

```ts
return originalWriteHead.apply(this, [statusCode, ...args] as [
  statusCode: number,
  headers?: OutgoingHttpHeaders | OutgoingHttpHeader[],
])
```

Is this better?

---

`return (originalWriteHead as any).apply(this, [statusCode, ...args]);`?

---

cc3435d04e0e2f3906b3742d7df54241c82c6aa0